File: src/models/Event.js
/**
 * Event Model
 * AI-Powered Logistics Monitoring System
 * 
 * Represents logistics tracking events with audit trail and encryption
 */

const mongoose = require('mongoose');
const crypto = require('crypto');

const eventSchema = new mongoose.Schema({
  eventId: {
    type: String,
    required: true,
    unique: true,
    index: true,
    default: () => crypto.randomUUID()
  },
  shipmentId: {
    type: String,
    required: true,
    index: true,
    ref: 'Shipment'
  },
  eventType: {
    type: String,
    required: true,
    enum: ['pickup', 'departure', 'arrival', 'delivery', 'delay', 'exception', 'customs_clearance', 'in_transit'],
    index: true
  },
  location: {
    address: { type: String, required: true },
    coordinates: {
      latitude: { type: Number, min: -90, max: 90 },
      longitude: { type: Number, min: -180, max: 180 }
    },
    locationCode: String,
    facilityType: String
  },
  timestamp: {
    type: Date,
    required: true,
    index: true
  },
  status: {
    type: String,
    required: true,
    enum: ['completed', 'in_progress', 'failed', 'pending']
  },
  description: {
    type: String,
    maxlength: 1000
  },
  sourceSystem: {
    type: String,
    required: true,
    enum: ['carrier_api', 'port_system', 'warehouse_system', 'customs_system', 'manual_entry', 'iot_sensor']
  },
  eventData: {
    type: Map,
    of: mongoose.Schema.Types.Mixed
  },
  // Security and audit fields
  createdBy: {
    type: String,
    required: true
  },
  verificationStatus: {
    type: String,
    enum: ['verified', 'unverified', 'disputed'],
    default: 'unverified'
  },
  verifiedBy: String,
  verifiedAt: Date,
  // Data lineage for compliance
  dataLineage: {
    sourceIp: String,
    sourceUser: String,
    sourceTimestamp: Date,
    processingTimestamp: Date,
    transformations: [String]
  },
  // Encryption metadata
  encryptionVersion: { type: String, default: 'v1' },
  checksumHash: String
}, {
  timestamps: true,
  collection: 'events'
});

// Compound indexes for query optimization
eventSchema.index({ shipmentId: 1, timestamp: -1 });
eventSchema.index({ eventType: 1, timestamp: -1 });
eventSchema.index({ sourceSystem: 1, createdAt: -1 });

// Pre-save middleware for checksum generation
eventSchema.pre('save', function(next) {
  // Generate checksum for data integrity
  const dataString = JSON.stringify({
    shipmentId: this.shipmentId,
    eventType: this.eventType,
    timestamp: this.timestamp,
    location: this.location
  });
  
  this.checksumHash = crypto.createHash('sha256').update(dataString).digest('hex');
  next();
});

// Method to verify event integrity
eventSchema.methods.verifyIntegrity = function() {
  const dataString = JSON.stringify({
    shipmentId: this.shipmentId,
    eventType: this.eventType,
    timestamp: this.timestamp,
    location: this.location
  });
  
  const currentHash = crypto.createHash('sha256').update(dataString).digest('hex');
  return currentHash === this.checksumHash;
};

// Static method for event aggregation
eventSchema.statics.getEventTimeline = async function(shipmentId) {
  return this.find({ shipmentId })
    .sort({ timestamp: 1 })
    .select('-eventData -dataLineage')
    .lean();
};

const Event = mongoose.model('Event', eventSchema);

module.exports = Event;

File: src/models/event.py
"""Logistics Event Model

Defines the LogisticsEvent entity for event monitoring.
"""

from datetime import datetime
from src.configs.database import db
from sqlalchemy.dialects.postgresql import UUID, JSONB
import uuid

class LogisticsEvent(db.Model):
    """Logistics event model for shipment tracking"""
    
    __tablename__ = 'logistics_events'
    
    id = db.Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    shipment_id = db.Column(UUID(as_uuid=True), db.ForeignKey('shipments.id'), nullable=False, index=True)
    
    # Event details
    event_type = db.Column(db.String(100), nullable=False)  # departure, arrival, delay, customs_clearance, delivery_attempt, exception
    event_description = db.Column(db.Text)
    event_timestamp = db.Column(db.DateTime, nullable=False, default=datetime.utcnow)
    
    # Location information
    location = db.Column(db.String(255))
    latitude = db.Column(db.Float)
    longitude = db.Column(db.Float)
    
    # Event severity and impact
    severity = db.Column(db.String(20), default='info')  # info, warning, error, critical
    impact_score = db.Column(db.Float)  # 0-100 scale
    
    # Source and attribution
    source_system = db.Column(db.String(100))  # API, sensor, manual, integration
    recorded_by = db.Column(UUID(as_uuid=True), db.ForeignKey('users.id'))
    
    # Processing status
    processed = db.Column(db.Boolean, default=False)
    processed_at = db.Column(db.DateTime)
    
    # Audit fields
    created_at = db.Column(db.DateTime, default=datetime.utcnow, nullable=False)
    
    # Metadata
    metadata = db.Column(JSONB, default={})
    
    def to_dict(self):
        """Convert model to dictionary"""
        return {
            'id': str(self.id),
            'shipment_id': str(self.shipment_id),
            'event_type': self.event_type,
            'event_description': self.event_description,
            'event_timestamp': self.event_timestamp.isoformat() + 'Z' if self.event_timestamp else None,
            'location': self.location,
            'latitude': self.latitude,
            'longitude': self.longitude,
            'severity': self.severity,
            'impact_score': self.impact_score,
            'source_system': self.source_system,
            'processed': self.processed,
            'processed_at': self.processed_at.isoformat() + 'Z' if self.processed_at else None,
            'created_at': self.created_at.isoformat() + 'Z' if self.created_at else None,
            'metadata': self.metadata
        }
    
    def __repr__(self):
        return f'<LogisticsEvent {self.event_type} for {self.shipment_id}>'